torch/csrc/jit/passes/create_autodiff_subgraphs.cpp

```diff
@@ -1,257 +1,293 @@
-    for(size_t i = 0; i < nn->outputs().size(); ++i) {
-      auto old_output = n->outputs()[i];
-      auto new_output = nn->outputs()[i];
-      value_map[old_output] = new_output;
-      std::vector<Use> to_replace;
-      for(auto u : old_output->uses()) {
-        // Uses within the set do not need to be made outputs
-        if(group_set.count(u.user) > 0)
-          continue;
-        // Other uses do, but we
-        // cannot replace them here or we invalid the uses list iterator
-        to_replace.push_back(u);
+
+    auto curNode = *block_->nodes().rbegin();
+    while (curNode != *block_->nodes().rend()) {
+      for (auto subBlock : curNode->blocks()) {
+        SubgraphSlicer(subBlock, minSubgraphSize_).run(diffGraphs);
       }
-      if(to_replace.size() > 0) {
-        new_graph->registerOutput(new_output);
-        Value * external_output = group_node->addOutput()->setType(old_output->type());
-        for(auto u : to_replace) {
-          u.user->replaceInput(u.offset, external_output);
+
+      // Save the previous node, since we might delete `curNode` in next block
+      auto prevNode = curNode->prev();
+      if (curNode->kind() == prim::DifferentiableGraph) {
+        // Inlining nodes may cause some subexpression to come back in the
+        // subgraphs (for example, copying constants in repeatedly will generate
+        // redundant prim::Constants). Run CSE to clean them up.
+        EliminateCommonSubexpression(*curNode->g(attr::Subgraph));
+
+        if (!inlineIfTooSmall(curNode)) {
+          diffGraphs.push_back(curNode);
         }
       }
+      curNode = prevNode;
     }
   }
-  group_node->insertBefore(nodes[0]);
-  // delete backward, so that nodes are use-free before deletion
-  for(size_t i = nodes.size(); i > 0; --i) {
-    nodes[i - 1]->destroy();
+
+ private:
+  static Graph& getSubgraph(Node* n) {
+    JIT_ASSERT(n->kind() == prim::DifferentiableGraph);
+    return *n->g(attr::Subgraph);
   }
-  JIT_ASSERT(isDifferentiable(*new_graph));
-  return group_node;
-}
 
-bool shouldConsiderForMerge(detail::Vertex<Node*>* v) {
-  if (v->data.size() >= 2) {
+  // Inline this node's group subgraph into the outer graph if it's smaller
+  // than the specified minimum size.
+  //
+  // Returns true if an inlining has occured, false otherwise.
+  bool inlineIfTooSmall(Node* n) {
+    JIT_ASSERT(n->kind() == prim::DifferentiableGraph);
+    auto& subgraph = getSubgraph(n);
+    size_t i = 0;
+    for (const auto node __attribute__((unused)) : subgraph.nodes()) {
+      if (++i >= minSubgraphSize_) {
+        return false;
+      }
+    }
+
+    unmergeGroup(n);
     return true;
   }
-  JIT_ASSERT(v->data.size() == 1);
-  auto * node = *v->data.begin();
-  if (node->kind() == prim::Constant) {
-    return false;
+
+  // Move nodes from producerGroup's subgraph to the top-level graph.
+  // This destroys `producerGroup`.
+  std::vector<Node*> unmergeGroup(Node* producerGroup) {
```

Comment:
afaik there are no explicit tests to check for the correctness of the inlining behavior; could you add one?